apps/meteor/ee/server/hooks/messages/BeforeSaveCannedResponse.ts

```diff
@@ -0,0 +1,116 @@
+import { isILivechatVisitor, isOmnichannelRoom } from '@rocket.chat/core-typings';
+import type { IMessage, IRoom, ISetting, IUser, SettingValue } from '@rocket.chat/core-typings';
+import { License } from '@rocket.chat/license';
+import { LivechatVisitors, Users } from '@rocket.chat/models';
+import get from 'lodash.get';
+import mem from 'mem';
+
+const placeholderFields = {
+	'contact.name': {
+		from: 'visitor',
+		dataKey: 'name',
+	},
+	'contact.email': {
+		from: 'visitor',
+		dataKey: 'visitorEmails[0].address',
+	},
+	'contact.phone': {
+		from: 'visitor',
+		dataKey: 'phone[0].phoneNumber',
+	},
+	'agent.name': {
+		from: 'agent',
+		dataKey: 'name',
+	},
+	'agent.email': {
+		from: 'agent',
+		dataKey: 'emails[0].address',
+	},
+};
+
+export class BeforeSaveCannedResponse {
```

Comment:
Should we have like a base class for these actions? 👀

Reply:
yes. I was thinking about this.. I couldn't think of a standard yet though